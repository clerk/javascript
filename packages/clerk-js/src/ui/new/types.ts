import type { SignInProps, SignUpProps } from '@clerk/types';
import type { ComponentType } from 'react';

export interface ComponentDefinition {
  type: 'component' | 'modal';
  load: () => Promise<{ default: ComponentType }>;
}

export type ClerkNewComponents = {
  SignIn: SignInProps;
  SignUp: SignUpProps;
};
