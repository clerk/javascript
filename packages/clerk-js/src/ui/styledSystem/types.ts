// eslint-disable-next-line no-restricted-imports
import { Interpolation as _Interpolation } from '@emotion/react';
import React from 'react';

import { InternalTheme } from '../foundations';

type StyleRule = Exclude<_Interpolation<InternalTheme>, string | number | boolean>;

/**
 * Primitives can override their styles using the css prop.
 * For customising layout/theme, prefer using the props defined on the component.
 */
type ThemableCssProp = ((params: InternalTheme) => StyleRule) | StyleRule;
type CssProp = { css?: ThemableCssProp };

export type AsProp = { as?: React.ElementType };

type ElementProps = {
  div: JSX.IntrinsicElements['div'];
  input: JSX.IntrinsicElements['input'];
  button: JSX.IntrinsicElements['button'];
  heading: JSX.IntrinsicElements['h1'];
  p: JSX.IntrinsicElements['p'];
  a: JSX.IntrinsicElements['a'];
  label: JSX.IntrinsicElements['label'];
  img: JSX.IntrinsicElements['img'];
  form: JSX.IntrinsicElements['form'];
};

/**
 * Some elements, like Flex can accept StateProps
 * simply because they need to be targettable when their container
 * component has a specific state. We then remove the props
 * before rendering the element to the DOM
 */
type StateProps = Partial<Record<'isDisabled' | 'hasError' | 'isLoading' | 'isOpen' | 'isActive', any>>;
/**
 * The form control elements can also accept a isRequired prop on top of the StateProps
 * We're handling it differently since this is a prop that cannot change - a required field
 * will remain required throughout the lifecycle of the component
 */
type RequiredProp = Partial<Record<'isRequired', boolean>>;

type PrimitiveProps<HtmlT extends keyof ElementProps> = ElementProps[HtmlT] & CssProp;
type PickSiblingProps<C extends React.FunctionComponent, T extends keyof Parameters<C>[0]> = Pick<Parameters<C>[0], T>;
type PropsOfComponent<C extends (...args: any[]) => any> = Parameters<C>[0];

export type {
  InternalTheme,
  PrimitiveProps,
  PickSiblingProps,
  PropsOfComponent,
  StyleRule,
  ThemableCssProp,
  StateProps,
  RequiredProp,
};
