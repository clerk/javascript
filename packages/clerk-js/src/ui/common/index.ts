export * from './authPropHelpers';
export * from './constants';
export * from './forms';
export * from './redirects';
export * from './verification';
export * from './withRedirectToHome';
export * from './SSOCallback';
export * from './EmailLinkVerify';
export * from './EmailLinkStatusCard';
export * from './Wizard';
export * from './PrintableComponent';
export * from './withOrganizationsEnabledGuard';
export * from './QRCode';
