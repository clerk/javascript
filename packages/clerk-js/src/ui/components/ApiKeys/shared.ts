import { useClerk } from '@clerk/shared/react';
import { useState } from 'react';

import { useFetch } from '../../hooks';
import type { Expiration } from './CreateApiKeyForm';

function getTimeLeftInSeconds(expirationOption: Expiration) {
  if (expirationOption === 'never') {
    return;
  }
  const now = new Date();
  const future = new Date(now);
  if (expirationOption === '30d') {
    future.setDate(future.getDate() + 30);
  } else if (expirationOption === '90d') {
    future.setDate(future.getDate() + 90);
  } else {
    throw new Error('TODO: Improve time handling');
  }
  const diffInMs = future.getTime() - now.getTime();
  const diffInSecs = Math.floor(diffInMs / 1000);
  return diffInSecs;
}

export function useApiKeys({ subject, perPage = 5 }: { subject: string; perPage?: number }) {
  const clerk = useClerk();
  const {
    data: apiKeys,
    isLoading,
    revalidate,
  } = useFetch(clerk.getApiKeys, { subject }, undefined, `api-key-source-${subject}`);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState('');
  const itemsPerPage = perPage;

  const filteredApiKeys = (apiKeys ?? []).filter(key => key.name.toLowerCase().includes(search.toLowerCase()));
  const itemCount = filteredApiKeys.length;
  const pageCount = Math.max(1, Math.ceil(itemCount / itemsPerPage));
  const startingRow = itemCount > 0 ? (page - 1) * itemsPerPage + 1 : 0;
  const endingRow = Math.min(page * itemsPerPage, itemCount);
  const paginatedApiKeys = filteredApiKeys.slice(startingRow - 1, endingRow);

  const handleCreate = async (params: {
    name: string;
    description?: string;
    expiration: Expiration;
    closeFn: () => void;
  }) => {
    await clerk.createApiKey({
      name: params.name,
      creationReason: params.description,
      secondsUntilExpiration: getTimeLeftInSeconds(params.expiration),
    });
    params.closeFn();
    revalidate();
  };

  const revokeApiKey = async (apiKeyID: string) => {
    await clerk.revokeApiKey({ apiKeyID, revocationReason: 'Revoked by user' });
    setPage(1);
    revalidate();
  };

  return {
    apiKeys: paginatedApiKeys,
    isLoading: isLoading ?? false,
    revokeApiKey,
    search,
    setSearch,
    page,
    setPage,
    pageCount,
    itemCount,
    startingRow,
    endingRow,
    handleCreate,
  };
}
