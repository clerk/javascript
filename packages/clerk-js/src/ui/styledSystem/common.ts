import { InternalTheme } from './types';

const textVariants = (t: InternalTheme) => {
  const base = {
    WebkitFontSmoothing: t.options.$fontSmoothing,
    fontFamily: 'inherit',
  };

  const smallRegular = {
    ...base,
    fontWeight: t.fontWeights.$normal,
    fontSize: t.fontSizes.$xs,
    lineHeight: t.lineHeights.$shorter,
  } as const;

  const smallMedium = {
    ...smallRegular,
    fontWeight: t.fontWeights.$medium,
    lineHeight: t.lineHeights.$short,
  } as const;

  const smallBold = {
    ...smallMedium,
    fontWeight: t.fontWeights.$bold,
  } as const;

  const extraSmallRegular = {
    ...base,
    fontWeight: t.fontWeights.$normal,
    fontSize: t.fontSizes.$2xs,
    letterSpacing: t.letterSpacings.$normal,
    lineHeight: t.lineHeights.$none,
  } as const;

  const regularRegular = {
    ...base,
    fontWeight: t.fontWeights.$normal,
    fontSize: t.fontSizes.$sm,
    lineHeight: t.lineHeights.$shorter,
  } as const;

  const regularMedium = {
    ...regularRegular,
    fontWeight: t.fontWeights.$medium,
  } as const;

  const largeBold = {
    ...base,
    fontWeight: t.fontWeights.$bold,
    fontSize: t.fontSizes.$md,
    lineHeight: t.lineHeights.$taller,
  } as const;

  const largeMedium = {
    ...largeBold,
    fontWeight: t.fontWeights.$medium,
  };

  const xlargeMedium = {
    ...largeBold,
    fontSize: t.fontSizes.$xl,
  } as const;

  const xxlargeMedium = {
    ...xlargeMedium,
    fontSize: t.fontSizes.$2xl,
  } as const;

  const buttonExtraSmallBold = {
    ...extraSmallRegular,
    fontWeight: t.fontWeights.$bold,
    textTransform: 'uppercase',
    fontFamily: t.fonts.$buttons,
  } as const;

  const buttonSmallRegular = {
    ...smallRegular,
    fontFamily: t.fonts.$buttons,
    lineHeight: t.lineHeights.$none,
  };

  const buttonRegularRegular = {
    ...regularRegular,
    fontFamily: t.fonts.$buttons,
    lineHeight: t.lineHeights.$none,
  };

  const buttonRegularMedium = {
    ...regularMedium,
    fontFamily: t.fonts.$buttons,
    lineHeight: t.lineHeights.$none,
  };

  return {
    buttonExtraSmallBold,
    buttonSmallRegular,
    buttonRegularRegular,
    buttonRegularMedium,
    extraSmallRegular,
    smallRegular,
    smallMedium,
    smallBold,
    regularRegular,
    regularMedium,
    largeMedium,
    largeBold,
    xlargeMedium,
    xxlargeMedium,
  } as const;
};

const fontSizeVariants = (t: InternalTheme) => {
  return {
    xss: { fontSize: t.fontSizes.$2xs },
    xs: { fontSize: t.fontSizes.$xs },
    sm: { fontSize: t.fontSizes.$sm },
  } as const;
};

const borderVariants = (t: InternalTheme, props?: any) => {
  return {
    normal: {
      borderRadius: t.radii.$md,
      border: t.borders.$normal,
      ...borderColor(t, props),
    },
  } as const;
};

const borderColor = (t: InternalTheme, props?: any) => {
  return {
    borderColor: props?.hasError ? t.colors.$danger500 : t.colors.$blackAlpha300,
  } as const;
};

const focusRing = (t: InternalTheme) => {
  return {
    '&:focus': {
      '&::-moz-focus-inner': { border: '0' },
      WebkitTapHighlightColor: 'transparent',
      boxShadow: t.shadows.$focusRing.replace('{{color}}', t.colors.$primary200),
      transitionProperty: t.transitionProperty.$common,
      transitionTimingFunction: t.transitionTiming.$common,
      transitionDuration: t.transitionDuration.$focusRing,
    },
  } as const;
};

const focusRingInput = (t: InternalTheme) => {
  return {
    '&:focus': {
      WebkitTapHighlightColor: 'transparent',
      boxShadow: t.shadows.$focusRingInput.replace('{{color}}', t.colors.$primary200),
      transitionProperty: t.transitionProperty.$common,
      transitionTimingFunction: t.transitionTiming.$common,
      transitionDuration: t.transitionDuration.$focusRing,
      borderColor: t.colors.$primary200,
    },
  } as const;
};

const disabled = (t: InternalTheme) => {
  return {
    '&:disabled,&[data-disabled]': {
      cursor: 'not-allowed',
      pointerEvents: 'none',
      opacity: t.opacity.$disabled,
    },
  } as const;
};

const centeredFlex = (display: 'flex' | 'inline-flex' = 'flex') => ({
  display: display,
  justifyContent: 'center',
  alignItems: 'center',
});

const maxHeightScroller = (t: InternalTheme) =>
  ({
    height: '100%',
    overflowY: 'scroll',
    '::-webkit-scrollbar': {
      background: 'transparent',
      width: '8px',
      height: '8px',
    },
    '::-webkit-scrollbar-thumb': {
      background: t.colors.$blackAlpha500,
    },
    '::-webkit-scrollbar-track': {
      background: 'transparent',
    },
  } as const);

export const common = {
  textVariants,
  fontSizeVariants,
  borderVariants,
  focusRing,
  focusRingInput,
  disabled,
  borderColor,
  centeredFlex,
  maxHeightScroller,
};
