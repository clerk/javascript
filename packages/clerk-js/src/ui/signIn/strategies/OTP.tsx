import { Label } from '@clerk/shared/components/label';
import {
  OneTimeCodeInput,
  VerifyCodeHandler,
} from '@clerk/shared/components/oneTimeCodeInput';
import { SignInFactor } from '@clerk/types';
import React from 'react';
import { FieldState } from 'ui/common';
import { useCoreSignIn } from 'ui/contexts';

export type OTPProps = {
  code: FieldState<string>;
  verifyCode: VerifyCodeHandler;
  factor: SignInFactor;
};

export function OTP({
  factor,
  code,
  verifyCode,
}: OTPProps): JSX.Element | null {
  const signIn = useCoreSignIn();

  if (factor.strategy !== 'email_code' && factor.strategy !== 'phone_code') {
    return null;
  }

  // TODO: https://www.notion.so/clerkdev/c8719edf0d5041e0b4d263a7ee574b7c
  const handleResentClick = () => {
    signIn.prepareFirstFactor(factor);
  };

  return (
    <>
      <Label className='cl-auth-form-message'>
        Enter the 6-digit code sent to
        <br />
        <strong>{factor.safeIdentifier}</strong>
      </Label>
      <OneTimeCodeInput
        value={code.value}
        setValue={code.setValue}
        verifyCodeHandler={verifyCode}
        onResendCode={handleResentClick}
        className='cl-otp-input'
      />
    </>
  );
}
