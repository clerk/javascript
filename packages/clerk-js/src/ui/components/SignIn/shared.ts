import { isClerkRuntimeError, isUserLockedError } from '@clerk/shared/error';
import { useClerk } from '@clerk/shared/react';
import { useCallback, useEffect } from 'react';

import { clerkInvalidFAPIResponse } from '../../../core/errors';
import { __internal_WebAuthnAbortService } from '../../../utils/passkeys';
import { useCoreSignIn, useSignInContext } from '../../contexts';
import { useCardState } from '../../elements';
import { useSupportEmail } from '../../hooks/useSupportEmail';
import { handleError } from '../../utils';

function useHandleAuthenticateWithPasskey(onSecondFactor: () => Promise<unknown>) {
  const card = useCardState();
  // @ts-expect-error -- private method for the time being
  const { setActive, __internal_navigateWithError } = useClerk();
  const supportEmail = useSupportEmail();
  const { afterSignInUrl } = useSignInContext();
  const { authenticateWithPasskey } = useCoreSignIn();

  useEffect(() => {
    return () => {
      __internal_WebAuthnAbortService.abort();
    };
  }, []);

  return useCallback(async (...args: Parameters<typeof authenticateWithPasskey>) => {
    try {
      const res = await authenticateWithPasskey(...args);
      switch (res.status) {
        case 'complete':
          return setActive({ session: res.createdSessionId, redirectUrl: afterSignInUrl });
        case 'needs_second_factor':
          return onSecondFactor();
        default:
          return console.error(clerkInvalidFAPIResponse(res.status, supportEmail));
      }
    } catch (err) {
      const { flow } = args[0] || {};

      if (isClerkRuntimeError(err)) {
        // In any case if the call gets aborted we should skip showing an error. This prevents updating the state of unmounted components.
        if (err.code === 'passkey_operation_aborted') {
          return;
        }
        // In case of autofill, if retrieval of credentials is cancelled by the user avoid showing errors as it results to pour UX.
        if (flow === 'autofill' && err.code === 'passkey_retrieval_cancelled') {
          return;
        }
      }

      if (isUserLockedError(err)) {
        return __internal_navigateWithError('..', err.errors[0]);
      }
      handleError(err, [], card.setError);
    }
  }, []);
}

export { useHandleAuthenticateWithPasskey };
