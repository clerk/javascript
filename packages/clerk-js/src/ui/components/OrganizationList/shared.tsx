import type { UserOrganizationInvitationResource } from '@clerk/types';
import type { PropsWithChildren } from 'react';
import { forwardRef } from 'react';

import { Box, Button, Col, descriptors, Flex, Spinner } from '../../customizables';
import { OrganizationPreview, PreviewButton } from '../../elements';
import { ArrowRightIcon } from '../../icons';
import type { ThemableCssProp } from '../../styledSystem';
import { common } from '../../styledSystem';

export const PreviewListItems = (props: PropsWithChildren) => {
  return (
    <Col
      elementDescriptor={descriptors.organizationListPreviewItems}
      sx={t => ({
        maxHeight: `calc(8 * ${t.sizes.$12})`,
        overflowY: 'auto',
        ...common.unstyledScrollbar(t),
      })}
    >
      {props.children}
    </Col>
  );
};

const sharedStyles: ThemableCssProp = t => ({
  height: t.space.$14,
  padding: `${t.space.$2} ${t.space.$8}`,
});

export const PreviewListItem = (
  props: PropsWithChildren<{
    organizationData: UserOrganizationInvitationResource['publicOrganizationData'];
  }>,
) => {
  return (
    <Flex
      align='center'
      gap={2}
      sx={[
        _ => ({
          minHeight: 'unset',
          justifyContent: 'space-between',
        }),
        sharedStyles,
      ]}
      elementDescriptor={descriptors.organizationListPreviewItem}
    >
      <OrganizationPreview
        elementId='organizationList'
        avatarSx={t => ({ width: t.sizes.$10, height: t.sizes.$10 })}
        mainIdentifierSx={t => ({
          fontSize: t.fontSizes.$xl,
          color: t.colors.$colorText,
        })}
        organization={props.organizationData}
      />
      {props.children}
    </Flex>
  );
};

export const PreviewListSpinner = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <Box
      ref={ref}
      sx={t => ({
        width: '100%',
        height: t.space.$12,
        position: 'relative',
      })}
    >
      <Box
        sx={{
          margin: 'auto',
          position: 'absolute',
          left: '50%',
          top: '50%',
          transform: 'translateY(-50%) translateX(-50%)',
        }}
      >
        <Spinner
          size='md'
          colorScheme='primary'
        />
      </Box>
    </Box>
  );
});

export const PreviewListItemButton = (props: Parameters<typeof Button>[0]) => {
  return (
    <Button
      elementDescriptor={descriptors.organizationListPreviewItemActionButton}
      textVariant='buttonExtraSmallBold'
      variant='outline'
      colorScheme='neutral'
      size='sm'
      {...props}
    />
  );
};

export const OrganizationListPreviewButton = (props: PropsWithChildren<{ onClick: () => void | Promise<void> }>) => {
  return (
    <PreviewButton
      elementDescriptor={descriptors.organizationListPreviewButton}
      sx={[sharedStyles]}
      icon={ArrowRightIcon}
      iconProps={{
        size: 'lg',
      }}
      showIconOnHover={false}
      {...props}
    />
  );
};
