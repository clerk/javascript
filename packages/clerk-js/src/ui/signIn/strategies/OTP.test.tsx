import { noop, render, renderJSON, waitFor } from '@clerk/shared/testUtils';
import { SignInFactor } from '@clerk/types';
import React from 'react';
import { FieldState } from 'ui/common';

import { OTP } from './OTP';

jest.mock('ui/contexts', () => {
  return {
    useCoreSignIn: jest.fn(),
  };
});

describe('<OTP/>', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  const code = {
    name: 'code',
    value: '123456',
    setValue: noop,
    error: undefined,
    setError: noop,
  } as FieldState<string>;

  const factor = {
    strategy: 'phone_code',
    safe_identifier: 'jdoe@example.com',
  } as SignInFactor;

  it('renders the OTP form component', async () => {
    const tree = renderJSON(
      <OTP verifyCode={(verify) => verify()} factor={factor} code={code} />
    );
    await jest.runAllTimers();
    expect(tree).toMatchSnapshot();
  });

  it('renders the OTP form, enters a password verifies the code', async () => {
    const mockVerify = jest.fn((verify) => verify());

    render(<OTP verifyCode={mockVerify} factor={factor} code={code} />);

    await waitFor(() => {
      expect(mockVerify).toHaveBeenCalledTimes(1);
      // TODO: Check for the value
      // expect(mockSubmit).toHaveBeenCalledWith({})
    });
  });
});
