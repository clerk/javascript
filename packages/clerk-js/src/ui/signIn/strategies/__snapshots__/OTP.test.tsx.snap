// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OTP/> renders the OTP form component 1`] = `
Array [
  <label
    className="label cl-auth-form-message"
  >
    Enter the 6-digit code sent to
    <br />
    <strong>
      jdoe@example.com
    </strong>
  </label>,
  <div
    className="cl-otp-input"
  >
    <div
      className="inputContainer"
      data-test-id="otp-container"
    >
      <input
        aria-label="Enter verification code. Digit 1"
        autoComplete="one-time-code"
        autoFocus={true}
        className="input codeInput"
        disabled={false}
        inputMode="numeric"
        maxLength={1}
        name="codeInput-0"
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onPaste={[Function]}
        type="text"
        value="1"
      />
      <input
        aria-label="Digit 2"
        autoComplete="one-time-code"
        autoFocus={false}
        className="input codeInput"
        disabled={false}
        inputMode="numeric"
        maxLength={1}
        name="codeInput-1"
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onPaste={[Function]}
        type="text"
        value="2"
      />
      <input
        aria-label="Digit 3"
        autoComplete="one-time-code"
        autoFocus={false}
        className="input codeInput"
        disabled={false}
        inputMode="numeric"
        maxLength={1}
        name="codeInput-2"
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onPaste={[Function]}
        type="text"
        value="3"
      />
      <input
        aria-label="Digit 4"
        autoComplete="one-time-code"
        autoFocus={false}
        className="input codeInput"
        disabled={false}
        inputMode="numeric"
        maxLength={1}
        name="codeInput-3"
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onPaste={[Function]}
        type="text"
        value="4"
      />
      <input
        aria-label="Digit 5"
        autoComplete="one-time-code"
        autoFocus={false}
        className="input codeInput"
        disabled={false}
        inputMode="numeric"
        maxLength={1}
        name="codeInput-4"
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onPaste={[Function]}
        type="text"
        value="5"
      />
      <input
        aria-label="Digit 6"
        autoComplete="one-time-code"
        autoFocus={false}
        className="input codeInput"
        disabled={false}
        inputMode="numeric"
        maxLength={1}
        name="codeInput-5"
        onChange={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onPaste={[Function]}
        type="text"
        value="6"
      />
    </div>
    <div
      className="statusContainer"
    >
      <button
        className="button link primary resendCode"
        disabled={false}
        onClick={[Function]}
      >
        Resend Code
      </button>
    </div>
  </div>,
]
`;
